component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -161,58 +160,191 @@ protected void initiateAuthenticationRequest(HttpServletRequest request,
+                        log.debug("Cannot find the authenticated user.");
+                    }
+                    throw new AuthenticationFailedException
+                            ("Authentication failed!. Cannot find the authenticated user.");
                 }
-                email = getEmailValueForUsername(username, context);
-                if (StringUtils.isEmpty(email)) {
-                    if (request.getParameter(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS) == null) {
-                        redirectToEmailAddressReqPage(response, context, emailOTPParameters, queryParams, username);
+
+                if (isLocalUser) {
+                    if (isEmailOTPDisableForUser(username, context, emailOTPParameters)) {
+                        if (isEmailOTPMandatory) {
+                            if (log.isDebugEnabled()) {
+                                log.debug("Email OTP is disabled for the user " + username + ", since email OTP is " +
+                                        "mandatory.");
+                            }
+                            String retryParam = EmailOTPAuthenticatorConstants.ERROR_EMAILOTP_DISABLE;
+                            redirectToErrorPage(response, context, emailOTPParameters, queryParams, retryParam);
+                        } else {
+                            // Skip email OTP step
+                            if (log.isDebugEnabled()) {
+                                log.debug("Email OTP step is skipped since email OTP is disabled for the user " +
+                                                username + "and email OTP is not mandatory.");
+                            }
+                            processFirstStepOnly(authenticatedUser, context);
+                        }
+
                     } else {
-                        updateEmailAddressForUsername(context, request, username);
                         email = getEmailValueForUsername(username, context);
+                        if (StringUtils.isEmpty(email)) {
+                            if (request.getParameter(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS) == null) {
+                                if (log.isDebugEnabled()) {
+                                    log.debug("email claim is not available for user " + username);
+                                }
+                                redirectToEmailAddressReqPage(response, context, emailOTPParameters, queryParams,
+                                        username);
+                            } else {
+                                updateEmailAddressForUsername(context, request, username);
+                                email = getEmailValueForUsername(username, context);
+                            }
+                        }
+                        if (StringUtils.isNotEmpty(email)) {
+                            processEmailOTPFlow(request, response, email, username, queryParams, context);
+                        }
```

Comment:
What should happen if this IF check fails?

Reply:
There should not be a case this if check gets fail because if the email is empty, then the above block will be executed and the user will be asked to enter an email address and the user attribute will be updated with the provided email address. But there is a chance if the user tries to update the email with an empty string, this check gets fail. I will correct that.